mypy/checker.py

```diff
@@ -3986,11 +3986,33 @@ def find_isinstance_check_helper(self, node: Expression) -> Tuple[TypeMap, TypeM
                 if len(node.args) != 2:  # the error will be reported elsewhere
                     return {}, {}
                 if literal(expr) == LITERAL_TYPE:
-                    return self.conditional_type_map_with_intersection(
+                    if_map, else_map = self.conditional_type_map_with_intersection(
                         expr,
                         type_map[expr],
                         get_isinstance_type(node.args[1], type_map),
                     )
+                    first_node_arg = node.args[0]
+                    if isinstance(first_node_arg, IndexExpr):
+                        arg_type = get_proper_type(type_map[first_node_arg.base])
+                        if isinstance(arg_type, UnionType) and if_map is not None and else_map is not None:
+                            var_name = first_node_arg.base
+                            if_branch_union = []
+                            else_branch_union = []
+                            t = if_map[expr]
+                            for x in arg_type.items:
+                                x = get_proper_type(x)
+                                if not isinstance(x, Instance) or not x.args:
+                                    return if_map, else_map
+                                if is_overlapping_types(x.args[0], t) is not covers_at_runtime(x.args[0], t, False):
```

Comment:
wasn't sure about this, essentially needed this special case to get `object` to work as expected in the test cases